File: app/src/main/java/com/hippo/ehviewer/dao/HistoryInfo.java
package com.hippo.ehviewer.dao;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.Index;

import android.os.Parcel;
import com.hippo.ehviewer.client.data.GalleryInfo;

/**
 * Entity mapped to table "HISTORY".
 * Primary key is GID (inherited from GalleryInfo).
 */
@Entity(tableName = "HISTORY", primaryKeys = {"GID"}, indices = {
    @Index("SERVER_PROFILE_ID"),
    @Index("TIME")
})
public class HistoryInfo extends GalleryInfo {

	@ColumnInfo(name = "MODE") public int mode;
	@ColumnInfo(name = "TIME") public long time;

	public static final Creator<HistoryInfo> CREATOR = new Creator<HistoryInfo>() {
		@Override
		public HistoryInfo createFromParcel(Parcel source) {
			return new HistoryInfo(source);
		}

		@Override
		public HistoryInfo[] newArray(int size) {
			return new HistoryInfo[size];
		}
	};

	public HistoryInfo() {
	}

	@Ignore
	public HistoryInfo(long gid) {
		this.gid = gid;
	}

	@Ignore
	public HistoryInfo(long gid, String token, String title, String titleJpn, String thumb, int category, String posted,
			String uploader, float rating, String simpleLanguage, int mode, long time) {
		this.gid = gid;
		this.token = token;
		this.title = title;
		this.titleJpn = titleJpn;
		this.thumb = thumb;
		this.category = category;
		this.posted = posted;
		this.uploader = uploader;
		this.rating = rating;
		this.simpleLanguage = simpleLanguage;
		this.mode = mode;
		this.time = time;
	}

	public long getGid() { return gid; }
	public void setGid(long gid) { this.gid = gid; }
	public String getToken() { return token; }
	public void setToken(String token) { this.token = token; }
	public String getTitle() { return title; }
	public void setTitle(String title) { this.title = title; }
	public String getTitleJpn() { return titleJpn; }
	public void setTitleJpn(String titleJpn) { this.titleJpn = titleJpn; }
	public String getThumb() { return thumb; }
	public void setThumb(String thumb) { this.thumb = thumb; }
	public int getCategory() { return category; }
	public void setCategory(int category) { this.category = category; }
	public String getPosted() { return posted; }
	public void setPosted(String posted) { this.posted = posted; }
	public String getUploader() { return uploader; }
	public void setUploader(String uploader) { this.uploader = uploader; }
	public float getRating() { return rating; }
	public void setRating(float rating) { this.rating = rating; }
	public String getSimpleLanguage() { return simpleLanguage; }
	public void setSimpleLanguage(String simpleLanguage) { this.simpleLanguage = simpleLanguage; }
	public int getMode() { return mode; }
	public void setMode(int mode) { this.mode = mode; }
	public long getTime() { return time; }
	public void setTime(long time) { this.time = time; }

	@Override
	public int describeContents() { return 0; }

	@Override
	public void writeToParcel(Parcel dest, int flags) {
		super.writeToParcel(dest, flags);
		dest.writeInt(this.mode);
		dest.writeLong(this.time);
	}

	@Ignore
	protected HistoryInfo(Parcel in) {
		super(in);
		this.mode = in.readInt();
		this.time = in.readLong();
	}

	@Ignore
	public HistoryInfo(GalleryInfo galleryInfo) {
		this.gid = galleryInfo.gid;
		this.token = galleryInfo.token;
		this.title = galleryInfo.title;
		this.titleJpn = galleryInfo.titleJpn;
		this.thumb = galleryInfo.thumb;
		this.category = galleryInfo.category;
		this.posted = galleryInfo.posted;
		this.uploader = galleryInfo.uploader;
		this.rating = galleryInfo.rating;
		this.simpleTags = galleryInfo.simpleTags;
		this.simpleLanguage = galleryInfo.simpleLanguage;
		this.serverProfileId = galleryInfo.serverProfileId;
	}

}

File: app/src/main/java/com/hippo/ehviewer/dao/HistoryInfo.kt
/*
 * Copyright 2017 Hippo Seven
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hippo.ehviewer.dao

import android.os.Parcel
import android.os.Parcelable

/**
 * In-memory view over the history subsystem of an `ARCHIVE_LOCAL_STATE`
 * row.
 *
 * Post-L1-4 this class is no longer a Room `@Entity` — the persisted
 * state lives on [ArchiveLocalState] (HISTORY_TIME / HISTORY_MODE
 * columns are non-null when the archive is in reading history). The
 * repository layer translates between this view and the unified row.
 */
class HistoryInfo() : Parcelable {

    @JvmField
    var arcid: String = ""

    @JvmField
    var title: String? = null

    @JvmField
    var thumb: String? = null

    @JvmField
    var rating: Float = 0f

    @JvmField
    var simpleLanguage: String? = null

    @JvmField
    var serverProfileId: Long = 0

    @JvmField
    var mode: Int = 0

    @JvmField
    var time: Long = 0

    /** Display tags; populated by the mapper from Archive.flatTags. */
    @JvmField
    var simpleTags: Array<String>? = null

    private constructor(`in`: Parcel) : this() {
        arcid = `in`.readString() ?: ""
        title = `in`.readString()
        thumb = `in`.readString()
        rating = `in`.readFloat()
        simpleLanguage = `in`.readString()
        serverProfileId = `in`.readLong()
        mode = `in`.readInt()
        time = `in`.readLong()
        simpleTags = `in`.createStringArray()
    }

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(arcid)
        dest.writeString(title)
        dest.writeString(thumb)
        dest.writeFloat(rating)
        dest.writeString(simpleLanguage)
        dest.writeLong(serverProfileId)
        dest.writeInt(mode)
        dest.writeLong(time)
        dest.writeStringArray(simpleTags)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<HistoryInfo> = object : Parcelable.Creator<HistoryInfo> {
            override fun createFromParcel(source: Parcel): HistoryInfo = HistoryInfo(source)
            override fun newArray(size: Int): Array<HistoryInfo?> = arrayOfNulls(size)
        }
    }
}
